testing/tests/test_runtime.py

```diff
@@ -153,13 +158,50 @@ def _on_start(self, _: ops.StartEvent):
             # simpler than causing the framework to raise it.
             raise _Abort(exit_code)
 
+    monkeypatch.setenv('SCENARIO_BARE_CHARM_ERRORS', bare_charm_errors)
     ctx = Context(MyCharm, meta={'name': 'foo'})
     if exit_code == 0:
         state_out = ctx.run(ctx.on.start(), State())
         assert {e.handle.kind for e in ctx.emitted_events} == {'start'}
         assert state_out.unit_status == ActiveStatus()
     else:
-        with pytest.raises(UncaughtCharmError) as exc:
+        error = _Abort if bare_charm_errors else UncaughtCharmError
+        with pytest.raises(error) as exc_info:
             ctx.run(ctx.on.start(), State())
-        assert isinstance(exc.value.__cause__, _Abort)
-        assert exc.value.__cause__.exit_code == exit_code
+        exc = exc_info.value if bare_charm_errors else exc_info.value.__cause__
+        assert isinstance(exc, _Abort)
+        assert exc.exit_code == exit_code
+
+
+class ValueErrorCharm(ops.CharmBase):
+    def __init__(self, framework: ops.Framework):
+        super().__init__(framework)
+        framework.observe(self.on.update_status, self._on_update_status)
+
+    def _on_update_status(self, event: ops.EventBase):
+        raise ValueError()
+
+
+@pytest.mark.parametrize(
+    ('expected_error', 'bare_charm_errors'),
+    (
+        (ValueError, '1'),
+        (ValueError, 'tRuE'),  # case insensitive
+        (UncaughtCharmError, ''),
+        (UncaughtCharmError, '✨ aNy ✩ sTrInG ✧ ReAlLy ...'),
```

Comment:
Nice, but maybe too clever.

How about:

```suggestion
        (ValueError, '1'),
        (ValueError, 'True'),
        (ValueError, 'true'),
        (UncaughtCharmError, '0'),
        (UncaughtCharmError, 'False'),
        (UncaughtCharmError, 'false'),
        (UncaughtCharmError, ''),
        (UncaughtCharmError, 'anything else'),
```

I would also swap them around (setting, observable_effect).

Aside: this doesn't test the missing env var case, though I imagine that's covered by regular tests.